### DISCUSSION TITLE
About automated-conversion, default arguments and everything related.

### DISCUSSION THREAD

**Discussion by thomasmansencal** [colour-science/colour#715] [Ideas]
@MichaelMauderer : So I don't want to step on your toes but I was poking at NetworkX a bit for some automated conversions to get a feel for it, nothing fancy but more as a test and to see what it would require.

So first a bit of simple code:

```python
import networkx as nx
import numpy as np

import colour

EDGES = [
    ['CIE XYZ', 'CIE xyY', colour.XYZ_to_xyY],
    ['CIE xyY', 'CIE XYZ', colour.xyY_to_XYZ],
    ['CIE XYZ', 'RGB', colour.XYZ_to_RGB],
    ['RGB', 'CIE XYZ', colour.RGB_to_XYZ],
    ['RGB', 'RGB', colour.RGB_to_RGB],
    ['RGB', 'HSV', colour.RGB_to_HSV],
    ['HSV', 'RGB', colour.HSV_to_RGB],
]


def _build_graph():
    graph = nx.DiGraph()

    for start, end, conversion_function in EDGES:
        graph.add_edge(start, end, conversion_function=conversion_function)

    return graph


GRAPH = _build_graph()


def conversion_path(start, end):
    path = nx.shortest_path(GRAPH, start, end)
    return [
        GRAPH.get_edge_data(a, b)['conversion_function']
        for a, b in zip(path[:-1], path[1:])
    ]


def convert(a, start, end, **kwargs):
    for conversion_function in conversion_path(start, end):
        print(conversion_function.__name__)
        a = conversion_function(a,
                                **colour.utilities.filter_kwargs(
                                    conversion_function, **kwargs))

    return a


print(convert(
    [0.5, 0.5, 0.5],
    'CIE XYZ',
    'HSV',
    illuminant_XYZ=np.array([0.34570, 0.35850]),
    illuminant_RGB=np.array([0.31270, 0.32900]),
    XYZ_to_RGB_matrix=np.array([
        [3.24062548, -1.53720797, -0.49862860],
        [-0.96893071, 1.87575606, 0.04151752],
        [0.05571012, -0.20402105, 1.05699594],
    ])))
```

So a few things came to mind when doing so:

-  First and not directly related: scaling as initially thought when working in the Metadata branch is actually a bit involved and I don't think we could have safely implemented that without the amount of work and testing done in the [Domain-Range Scale Support PR](https://github.com/colour-science/colour/pull/402).
- If we want to keep things simple we will probably have to force the [Domain-Range Scale to *1*](http://colour.readthedocs.io/en/feature-domain_range_scale/basics.html#scale-1), this comes with the annoyance that the output of some paths will not match the default *Reference* scale output. Alternatively we can painfully make the conversion required if we want to work in the *Reference* scale.
- It is obvious given the above call, that some definitions need default arguments, e.g. `colour.XYZ_to_RGB`. What will they be?
- As we define the default arguments I think we should maybe look at changing the default D50 illuminant to D65 for definitions that use it.

Thoughts?

**Comment by thomasmansencal** [comment]
Coincidentally, I just came across this in regard to CIE D Illuminant usage [1]:

> It is recommended that, in the interest of standardization, D65 be used whenever possible. When D65 cannot be used, it is recommended that one of the daylight illuminants D50, D55, or D75 defined in Table T.1 be used.

References
---
1. CIE TC 1-48. (2004). CIE 015:2004 Colorimetry, 3rd Edition. CIE 015:2004 Colorimetry, 3rd Edition. Commission internationale de l’éclairage. ISBN:978-3-901-90633-6

**Comment by MichaelMauderer** [comment]
No toe stepping. This looks very good.

> If we want to keep things simple we will probably have to force the Domain-Range Scale to 1, this comes with the annoyance that the output of some paths will not match the default Reference scale output. Alternatively, we can painfully make the conversion required if we want to work in the Reference scale.

Yes, internally it would be best to be consistent about the domain ranges that are expected and produced. But the final output could be rescaled to whatever the caller of the conversion wants.  

>   It is obvious given the above call, that some definitions need default arguments, e.g. colour.XYZ_to_RGB. What will they be?

We might also require those to be fed into the initial call and fail if they are not provided. (assuming no sensible default can be found for something)

>  As we define the default arguments I think we should maybe look at changing the default D50 illuminant to D65 for definitions that use it.

We just need to make sure no nasty surprises happen to existing code…

**Comment by thomasmansencal** [comment]
> (assuming no sensible default can be found for something)

I was leaning toward using still lowest common denominator, AKA sRGB.

> We just need to make sure no nasty surprises happen to existing code…

I think this should be visible very quickly through the tests, we will have to update a few of them actually, so it should give us a good overview. Whenever we make that change, it will be done through a singular atomic PR to assess the extent.

Did you have time to look at #402?

**Comment by thomasmansencal** [comment]
I have implemented a fair bunch of support for that in the [graph](https://github.com/colour-science/colour/tree/feature/graph) branch.

![image](https://user-images.githubusercontent.com/99779/63645365-c51f8880-c750-11e9-8f68-064758b27a66.png)
